### PR TITLE
Metrics plugin to support system and application error rates

### PR DESCRIPTION
**Pull request by sreevatsanraman**
CDAP Dashboard needs to surface System and application error rates (https://issues.cask.co/browse/CDAP-2079). The approach for getting the metrics involves extracting the log level from the logging context in the log-saver system component and emitting the metrics for various log levels. This task is broken down into two stages

- [x] Refactor Log Saver to have pluggable KafkaLogProcessors that can perform the two tasks (existing LogWriter functionality, new metrics capability) and in the future can be extended to do any additional capability
- [x] Provide the existing log writer functionality in the new framework
- [ ] Add the metrics plugin to emit metrics (upcoming PR)

Build: http://builds.cask.co/browse/CDAP-DUT1400-5

### CODE REVIEW

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/guice/LoggingModules.java [hunk 1]
```diff
@@ -44,6 +48,9 @@ protected void configure() {
         bind(LogAppender.class).annotatedWith(Names.named(LoggingConfiguration.SYNC_LOG_APPENDER_ANNOTATION))
           .to(FileLogAppender.class).in(Scopes.SINGLETON);
         bind(LogAppender.class).to(AsyncLogAppender.class).in(Scopes.SINGLETON);
+        Multibinder<KafkaLogProcessor> handlerBinder = Multibinder.newSetBinder
+          (binder(), KafkaLogProcessor.class, Names.named(Constants.LogSaver.MESSAGE_PROCESSORS));
+        handlerBinder.addBinding().to(KafkaLogWriterPlugin.class);
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/guice/LoggingModules.java]: We don't run LogSaver in in-memory mode, so this can be removed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/guice/LoggingModules.java [hunk 2]
```diff
@@ -57,6 +64,9 @@ protected void configure() {
...
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/guice/LoggingModules.java]: We don't run LogSaver in standalone mode, so this can be removed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+   *
+   * @param event instance of {@link KafkaLogEvent}
+   */
+  public void process(KafkaLogEvent event);
+
+  /**
+   * Called to perform cleanup tasks. This method will be called before the framework stops calling the process
+   * method any further.
+   */
+  public void cleanup();
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java]: `cleanup` does not imply that this processor will be stopped. Maybe call it `stop`?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        List<KafkaLogEvent> msgList = null;
+        if (entry == null) {
+          long eventArrivalBucketKey = System.currentTimeMillis() / eventBucketIntervalMs;
+          msgList = Lists.newArrayList();
+          messageTable.put(key, loggingContext.getLogPathFragment(logBaseDir),
+                           new AbstractMap.SimpleEntry<Long, List<KafkaLogEvent>>(eventArrivalBucketKey, msgList));
+        } else {
+          msgList = messageTable.get(key, loggingContext.getLogPathFragment(logBaseDir)).getValue();
+        }
+        LOG.info("ADDING event {} {}", event.getLogEvent().getMessage(), event.getLoggingContext());
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java]: This message can be removed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    messageTable.clear();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    // No-op
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java]: I don't see `shutDown` being called anywhere.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Kafka callback to fetch log messages and store them in time buckets per logging context.
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java]: Javadoc need updating.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java [hunk 7]
```diff
@@ -204,10 +143,6 @@ private void scheduleTasks(Set<Integer> partitions) throws Exception {
 
     subscribe(partitions);
 
-    LogWriter logWriter = new LogWriter(logFileWriter, messageTable,
-                                        eventBucketIntervalMs, maxNumberOfBucketsInTable);
-    logWriterFuture = scheduledExecutor.scheduleWithFixedDelay(logWriter, 100, 200, TimeUnit.MILLISECONDS);
-
     if (partitions.contains(0)) {
       LOG.info("Scheduling cleanup task");
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java]: Cleanup task has to be now moved to `KafkaLogWriterPlugin`, right?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+import java.util.SortedSet;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Plugin that writes the log data.
+ */
+public class KafkaLogWriterPlugin extends AbstractIdleService implements KafkaLogProcessor {
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java]: If `KafkaLogWriterPlugin` extends `AbstractIdleService`, then we can call `stop` instead of `cleanUp`, right?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java (hunk 3 above)
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java]: Also, javadoc says `This method will be called before the framework stops calling the process method any further`, but the processor plugin is reused across partition changes, and cleanup is called every time partition changes.

**Comment by poornachandra** [comment]
@sreevatsanraman I'm done with one pass of review.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java (hunk 4 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java]: Done.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java (hunk 7 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java]: Fixe.d

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java (hunk 6 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java]: Fixed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java (hunk 8 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java]: I fixed it. the plugin doesn't extend from `AbstractIdleService`. Instead uses `stop()` method to shutdown executor service.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java (hunk 5 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java]: Fixed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java (hunk 3 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java]: Updated.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/guice/LoggingModules.java (hunk 1 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/guice/LoggingModules.java]: Removed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+ * Kafka Message callback that fetches log from kafka and calls the process method for all the plugins configured.
+ */
+public class KafkaMessageCallback implements KafkaConsumer.MessageCallback {
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaMessageCallback.class);
+
+  private final Set<KafkaLogProcessor> kafkaLogProcessors;
+  private final LoggingEventSerializer serializer;
+  private final CountDownLatch countDownLatch;
+
+  public KafkaMessageCallback(CountDownLatch countDownLatch,
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java]: Maybe call it `stopLatch` for clarity?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+      Thread.currentThread().interrupt();
+      return;
+    }
+
+    int count = 0;
+
+    while (messages.hasNext()) {
+      FetchedMessage message = messages.next();
+      GenericRecord genericRecord = serializer.toGenericRecord(message.getPayload());
+      ILoggingEvent event = serializer.fromGenericRecord(genericRecord);
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java]: It would be a good idea to enclose the above de-serializing in try-catch so that one badly formatted log message does not impact all messages in this run.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+      LoggingContext loggingContext = LoggingContextHelper.getLoggingContext(event.getMDCPropertyMap());
+      KafkaLogEvent logEvent = new KafkaLogEvent(genericRecord, event, loggingContext,
+                                                 message.getTopicPartition().getPartition(),
+                                                 message.getNextOffset());
+
+      for (KafkaLogProcessor processor : kafkaLogProcessors) {
+        processor.process(logEvent);
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java]: Same thing here - good to enclose `process` in try-catch.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java (hunk 9 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java]: Fixed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java (hunk 10 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java]: Good point, fixed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java (hunk 11 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java]: Fixed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java [hunk 12]
```diff
@@ -61,110 +47,32 @@
...
   private Map<Integer, Cancellable> kafkaCancelMap;
   private Map<Integer, CountDownLatch> kafkaCancelCallbackLatchMap;
+  private Set<KafkaLogProcessor> messageProcessors;
+
 
   @Inject
   public LogSaver(CConfiguration cConf, CheckpointManager checkpointManager,
                   FileMetaDataManager fileMetaDataManager, KafkaClientService kafkaClient,
-                  CConfiguration cConfig, LocationFactory locationFactory) throws Exception {
+                  CConfiguration cConfig, LocationFactory locationFactory,
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java]: `cConf`, `fileMetaDataManager`, `cConfig` and `locationFactory` are unused.

Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java]: Fixed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+package co.cask.cdap.logging.save;
+
+import co.cask.cdap.logging.kafka.KafkaLogEvent;
+
+import java.util.Set;
+
+/**
+ * Process {@link KafkaLogEvent} with KafkaLogProcessor. Extends PartitionChangeHandler to implement the necessary
+ * changes for partitions changed events.
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java]: The second sentence from the javadoc can be removed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+  /**
+   * Process method will be called for each event received from Kafka from the topics published for log saver.
+   *
+   * @param event instance of {@link KafkaLogEvent}
+   */
+  public void process(KafkaLogEvent event);
+
+  /**
+   * Called to perform cleanup tasks. This method will be called partitionChanged is called as well as during shutdown.
+   * method any further.
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java]: This javadoc needs to be re-written.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+
+  @Override
+  public void init(Set<Integer> partitions) {
+
+    scheduledExecutor = MoreExecutors.listeningDecorator(Executors.newSingleThreadScheduledExecutor(
+      Threads.createDaemonThreadFactory("log-saver-log-processor")));;
+
+    LogWriter logWriter = new LogWriter(logFileWriter, messageTable,
+                                        eventBucketIntervalMs, maxNumberOfBucketsInTable);
+    //TODO: Is this a good idea?
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java]: TODO needs addressing.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+      if (cleanupFuture != null && !cleanupFuture.isCancelled() && !cleanupFuture.isDone()) {
+        cleanupFuture.cancel(false);
+        cleanupFuture = null;
+      }
+
+      countDownLatch.countDown();
+
+      if (scheduledExecutor != null) {
+        scheduledExecutor.shutdown();
+      }
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java]: Although this is not part of this PR, I would suggest to rearrange the stop logic to make it more resilient.
```java
if (countDownLatch != null) {
  countDownLatch.countDown();
}

if (scheduledExecutor != null) {
  scheduledExecutor.shutdown();
}
scheduledExecutor.awaitTermination(5, TimeUnit.MINUTES);

logFileWriter.flush();
logFileWriter.close();
```

We don't need `logWriterFuture` and `cleanupFuture` in this case.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+
+        LoggingContext loggingContext = LoggingContextHelper.getLoggingContext(event.getMDCPropertyMap());
+        KafkaLogEvent logEvent = new KafkaLogEvent(genericRecord, event, loggingContext,
+                                                   message.getTopicPartition().getPartition(),
+                                                   message.getNextOffset());
+
+        for (KafkaLogProcessor processor : kafkaLogProcessors) {
+          processor.process(logEvent);
+        }
+      } catch (Throwable th) {
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java]: It would be a better idea to catch process exception separately. This will allow us to pass the event to other processors when one processor throws an exception.

**Comment by poornachandra** [comment]
@sreevatsanraman Final few comments.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java (hunk 13 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java]: Fixed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java (hunk 14 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogProcessor.java]: Fixed

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java (hunk 15 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java]: Todo was a note to myself. Figured it out and removed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java (hunk 16 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java]: Fixed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java (hunk 17 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaMessageCallback.java]: Fixed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java [hunk 18]
```diff
@@ -201,46 +99,25 @@ private void scheduleTasks(Set<Integer> partitions) throws Exception {
...
-    if (logWriterFuture != null && !logWriterFuture.isCancelled() && !logWriterFuture.isDone()) {
-      logWriterFuture.cancel(false);
-      logWriterFuture = null;
-    }
-
-    if (cleanupFuture != null && !cleanupFuture.isCancelled() && !cleanupFuture.isDone()) {
-      cleanupFuture.cancel(false);
-      cleanupFuture = null;
+    for (KafkaLogProcessor processor : messageProcessors) {
+      processor.stop();
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java]: Would be a good idea to catch exception from stop here.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java [hunk 19]
```diff
@@ -248,6 +125,10 @@ private void cancelLogCollectorCallbacks() {
   private void subscribe(Set<Integer> partitions) throws Exception {
     LOG.info("Prepare to subscribe for partitions: {}", partitions);
 
+    for (KafkaLogProcessor processor : messageProcessors) {
+      processor.init(partitions);
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java]: Would be a good idea to catch exception from init here.

**Comment by poornachandra** [comment]
@sreevatsanraman Two minor comments, please fix and merge.

Would be a good idea to squash the commits if possible before merging.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java (hunk 18 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java]: Fixed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java (hunk 19 above)
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java]: Not fixing this, we should not process if there an exception on init for all plugins. Would give the admins a chance to look at the failure and not fail silently.
